File: android/src/main/kotlin/com/weilan/wechat_bridge/WechatBridgePlugin.kt
package com.weilan.wechat_bridge

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.content.pm.ProviderInfo
import android.net.Uri
import android.os.Build
import com.tencent.mm.opensdk.constants.Build as WxBuild
import com.tencent.mm.opensdk.diffdev.DiffDevOAuthFactory
import com.tencent.mm.opensdk.diffdev.IDiffDevOAuth
import com.tencent.mm.opensdk.diffdev.OAuthErrCode
import com.tencent.mm.opensdk.diffdev.OAuthListener
import com.tencent.mm.opensdk.modelbase.BaseResp
import com.tencent.mm.opensdk.modelbiz.OpenRankList
import com.tencent.mm.opensdk.modelbiz.OpenWebview
import com.tencent.mm.opensdk.modelbiz.SubscribeMessage
import com.tencent.mm.opensdk.modelbiz.WXLaunchMiniProgram
import com.tencent.mm.opensdk.modelbiz.WXOpenBusinessView
import com.tencent.mm.opensdk.modelbiz.WXOpenBusinessWebview
import com.tencent.mm.opensdk.modelbiz.WXOpenCustomerServiceChat
import com.tencent.mm.opensdk.modelmsg.LaunchFromWX
import com.tencent.mm.opensdk.modelmsg.SendAuth
import com.tencent.mm.opensdk.modelmsg.SendMessageToWX
import com.tencent.mm.opensdk.modelmsg.ShowMessageFromWX
import com.tencent.mm.opensdk.modelmsg.WXEmojiObject
import com.tencent.mm.opensdk.modelmsg.WXFileObject
import com.tencent.mm.opensdk.modelmsg.WXImageObject
import com.tencent.mm.opensdk.modelmsg.WXMediaMessage
import com.tencent.mm.opensdk.modelmsg.WXMiniProgramObject
import com.tencent.mm.opensdk.modelmsg.WXMusicObject
import com.tencent.mm.opensdk.modelmsg.WXTextObject
import com.tencent.mm.opensdk.modelmsg.WXVideoObject
import com.tencent.mm.opensdk.modelmsg.WXWebpageObject
import com.tencent.mm.opensdk.modelpay.PayReq
import com.tencent.mm.opensdk.modelpay.PayResp
import com.tencent.mm.opensdk.openapi.IWXAPI
import com.tencent.mm.opensdk.openapi.IWXAPIEventHandler
import com.tencent.mm.opensdk.openapi.WXAPIFactory
import androidx.core.content.FileProvider
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import java.io.File
import java.util.concurrent.atomic.AtomicBoolean

class WechatBridgePlugin : FlutterPlugin, ActivityAware, PluginRegistry.NewIntentListener, MethodChannel.MethodCallHandler {

    private var channel: MethodChannel? = null
    private var applicationContext: Context? = null
    private var activityPluginBinding: ActivityPluginBinding? = null

    private val qrauth: IDiffDevOAuth = DiffDevOAuthFactory.getDiffDevOAuth()
    private var iwxapi: IWXAPI? = null
    private val handleInitialWXReqFlag = AtomicBoolean(false)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "wechat_bridge")
        channel!!.setMethodCallHandler(this)
        applicationContext = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        channel = null
        applicationContext = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityPluginBinding = binding
        activityPluginBinding!!.addOnNewIntentListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        qrauth.removeAllListeners()
        activityPluginBinding = null
    }

    override fun onNewIntent(intent: Intent): Boolean {
        val extra = WechatCallbackActivity.extraCallback(intent)
        if (extra != null) {
            iwxapi?.handleIntent(extra, iwxapiEventHandler)
            return true
        }
        return false
    }

    private val iwxapiEventHandler = object : IWXAPIEventHandler {
        override fun onReq(req: com.tencent.mm.opensdk.modelbase.BaseReq) {
            val map = mutableMapOf<String, Any?>()
            when (req) {
                is LaunchFromWX.Req -> {
                    map["messageAction"] = req.messageAction
                    map["messageExt"] = req.messageExt
                    map["lang"] = req.lang
                    map["country"] = req.country
                    channel?.invokeMethod("onLaunchFromWXReq", map)
                }
                is ShowMessageFromWX.Req -> {
                    map["messageAction"] = req.message.messageAction
                    map["messageExt"] = req.message.messageExt
                    map["lang"] = req.lang
                    map["country"] = req.country
                    channel?.invokeMethod("onShowMessageFromWXReq", map)
                }
            }
        }

        override fun onResp(resp: BaseResp) {
            val map = mutableMapOf<String, Any?>(
                "errorCode" to resp.errCode,
                "errorMsg" to resp.errStr
            )
            when (resp) {
                is SendAuth.Resp -> {
                    if (resp.errCode == BaseResp.ErrCode.ERR_OK) {
                        map["code"] = resp.code
                        map["state"] = resp.state
                        map["lang"] = resp.lang
                        map["country"] = resp.country
                    }
                    channel?.invokeMethod("onAuthResp", map)
                }
                is OpenWebview.Resp -> channel?.invokeMethod("onOpenUrlResp", map)
                is SendMessageToWX.Resp -> channel?.invokeMethod("onShareMsgResp", map)
                is SubscribeMessage.Resp -> {
                    if (resp.errCode == BaseResp.ErrCode.ERR_OK) {
                        map["openId"] = resp.openId
                        map["templateId"] = resp.templateID
                        map["scene"] = resp.scene
                        map["action"] = resp.action
                        map["reserved"] = resp.reserved
                    }
                    channel?.invokeMethod("onSubscribeMsgResp", map)
                }
                is WXLaunchMiniProgram.Resp -> {
                    if (resp.errCode == BaseResp.ErrCode.ERR_OK) {
                        map["extMsg"] = resp.extMsg
                    }
                    channel?.invokeMethod("onLaunchMiniProgramResp", map)
                }
                is WXOpenCustomerServiceChat.Resp ->
                    channel?.invokeMethod("onOpenCustomerServiceChatResp", map)
                is WXOpenBusinessView.Resp -> {
                    if (resp.errCode == BaseResp.ErrCode.ERR_OK) {
                        map["businessType"] = resp.businessType
                        map["extMsg"] = resp.extMsg
                    }
                    channel?.invokeMethod("onOpenBusinessViewResp", map)
                }
                is WXOpenBusinessWebview.Resp -> {
                    if (resp.errCode == BaseResp.ErrCode.ERR_OK) {
                        map["businessType"] = resp.businessType
                        map["resultInfo"] = resp.resultInfo
                    }
                    channel?.invokeMethod("onOpenBusinessWebviewResp", map)
                }
                is PayResp -> {
                    if (resp.errCode == BaseResp.ErrCode.ERR_OK) {
                        map["returnKey"] = resp.returnKey
                    }
                    channel?.invokeMethod("onPayResp", map)
                }
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "registerApp" -> registerApp(call, result)
            "handleInitialWXReq" -> handleInitialWXReq(call, result)
            "isInstalled" -> result.success(iwxapi != null && iwxapi!!.isWXAppInstalled)
            "isSupportApi" -> result.success(
                iwxapi != null && iwxapi!!.wxAppSupportAPI >= WxBuild.OPENID_SUPPORTED_SDK_INT
            )
            "isSupportStateApi" -> result.success(
                iwxapi != null && iwxapi!!.wxAppSupportAPI >= WxBuild.SUPPORTED_SEND_TO_STATUS
            )
            "openWechat" -> result.success(iwxapi?.openWXApp() ?: false)
            "auth" -> handleAuthCall(call, result)
            "startQrauth", "stopQrauth" -> handleQRAuthCall(call, result)
            "openUrl" -> handleOpenUrlCall(call, result)
            "openRankList" -> handleOpenRankListCall(call, result)
            "shareText" -> handleShareTextCall(call, result)
            "shareImage", "shareFile", "shareEmoji", "shareMusic", "shareVideo",
            "shareWebpage", "shareMiniProgram" -> handleShareMediaCall(call, result)
            "subscribeMsg" -> handleSubscribeMsgCall(call, result)
            "launchMiniProgram" -> handleLaunchMiniProgramCall(call, result)
            "openCustomerServiceChat" -> handleOpenCustomerServiceChat(call, result)
            "openBusinessView" -> handleOpenBusinessView(call, result)
            "openBusinessWebview" -> handleOpenBusinessWebview(call, result)
            "pay" -> handlePayCall(call, result)
            else -> result.notImplemented()
        }
    }

    private fun registerApp(call: MethodCall, result: Result) {
        val appId = call.argument<String>("appId") ?: run {
            result.error("INVALID_ARGUMENT", "appId is required", null)
            return
        }
        iwxapi = WXAPIFactory.createWXAPI(applicationContext, appId)
        iwxapi!!.registerApp(appId)
        result.success(null)
    }

    private fun handleInitialWXReq(call: MethodCall, result: Result) {
        if (handleInitialWXReqFlag.compareAndSet(false, true)) {
            val activity = activityPluginBinding?.activity
            if (activity != null) {
                val extra = WechatCallbackActivity.extraCallback(activity.intent)
                if (extra != null) {
                    iwxapi?.handleIntent(extra, iwxapiEventHandler)
                }
            }
        }
        // 幂等：第二次及后续调用直接返回成功，避免 PlatformException(FAILED)
        result.success(null)
    }

    private fun handleAuthCall(call: MethodCall, result: Result) {
        val req = SendAuth.Req()
        req.scope = call.argument("scope")
        req.state = call.argument("state")
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleQRAuthCall(call: MethodCall, result: Result) {
        when (call.method) {
            "startQrauth" -> {
                val appId = call.argument<String>("appId")!!
                val scope = call.argument<String>("scope")!!
                val noncestr = call.argument<String>("noncestr")!!
                val timestamp = call.argument<String>("timestamp")!!
                val signature = call.argument<String>("signature")!!
                qrauth.auth(appId, scope, noncestr, timestamp, signature, qrauthListener)
            }
            "stopQrauth" -> qrauth.stopAuth()
        }
        result.success(null)
    }

    private val qrauthListener = object : OAuthListener {
        override fun onAuthGotQrcode(qrcodeImgPath: String?, imgBuf: ByteArray?) {
            imgBuf?.let {
                channel?.invokeMethod("onAuthGotQrcode", mapOf("imageData" to it))
            }
        }

        override fun onQrcodeScanned() {
            channel?.invokeMethod("onAuthQrcodeScanned", null)
        }

        override fun onAuthFinish(errCode: OAuthErrCode?, authCode: String?) {
            channel?.invokeMethod("onAuthFinish", mapOf(
                "errorCode" to (errCode?.code ?: 0),
                "authCode" to (authCode ?: "")
            ))
        }
    }

    private fun handleOpenUrlCall(call: MethodCall, result: Result) {
        val req = OpenWebview.Req()
        req.url = call.argument("url")
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleOpenRankListCall(call: MethodCall, result: Result) {
        val req = OpenRankList.Req()
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleShareTextCall(call: MethodCall, result: Result) {
        val req = SendMessageToWX.Req()
        req.transaction = "${call.method}: ${System.currentTimeMillis()}"
        req.scene = call.argument("scene") ?: 0
        val text = call.argument<String>("text") ?: ""
        val message = WXMediaMessage()
        message.description = text.take(1024)
        val obj = WXTextObject()
        obj.text = text
        message.mediaObject = obj
        req.message = message
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleShareMediaCall(call: MethodCall, result: Result) {
        val req = SendMessageToWX.Req()
        req.transaction = "${call.method}: ${System.currentTimeMillis()}"
        req.scene = call.argument<Int>("scene") ?: 0
        val message = WXMediaMessage()
        message.title = call.argument("title")
        message.description = call.argument("description")
        message.thumbData = call.argument("thumbData")

        when (call.method) {
            "shareImage" -> {
                val obj = WXImageObject()
                call.argument<ByteArray>("imageData")?.let { obj.imageData = it }
                call.argument<String>("imageUri")?.let { obj.imagePath = getShareFilePath(it) }
                message.mediaObject = obj
            }
            "shareFile" -> {
                val obj = WXFileObject()
                call.argument<ByteArray>("fileData")?.let { obj.fileData = it }
                call.argument<String>("fileUri")?.let { obj.filePath = getShareFilePath(it) }
                message.mediaObject = obj
            }
            "shareEmoji" -> {
                val obj = WXEmojiObject()
                call.argument<ByteArray>("emojiData")?.let { obj.emojiData = it }
                call.argument<String>("emojiUri")?.let { obj.emojiPath = getShareFilePath(it) }
                message.mediaObject = obj
            }
            "shareMusic" -> {
                val obj = WXMusicObject()
                obj.musicUrl = call.argument("musicUrl")
                obj.musicDataUrl = call.argument("musicDataUrl")
                obj.musicLowBandUrl = call.argument("musicLowBandUrl")
                obj.musicLowBandDataUrl = call.argument("musicLowBandDataUrl")
                message.mediaObject = obj
            }
            "shareVideo" -> {
                val obj = WXVideoObject()
                obj.videoUrl = call.argument("videoUrl")
                obj.videoLowBandUrl = call.argument("videoLowBandUrl")
                message.mediaObject = obj
            }
            "shareWebpage" -> {
                val obj = WXWebpageObject()
                obj.webpageUrl = call.argument("webpageUrl")
                message.mediaObject = obj
            }
            "shareMiniProgram" -> {
                val obj = WXMiniProgramObject()
                obj.webpageUrl = call.argument("webpageUrl")!!
                obj.userName = call.argument("username")!!
                obj.path = call.argument("path")
                call.argument<ByteArray>("hdImageData")?.let { message.thumbData = it }
                obj.withShareTicket = call.argument("withShareTicket") ?: false
                obj.miniprogramType = call.argument("type") ?: 0
                obj.disableforward = if (call.argument<Boolean>("disableForward") == true) 1 else 0
                message.mediaObject = obj
            }
        }
        req.message = message
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleSubscribeMsgCall(call: MethodCall, result: Result) {
        val req = SubscribeMessage.Req()
        req.scene = call.argument("scene") ?: 0
        req.templateID = call.argument("templateId")
        req.reserved = call.argument("reserved")
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleLaunchMiniProgramCall(call: MethodCall, result: Result) {
        val req = WXLaunchMiniProgram.Req()
        req.userName = call.argument("username")!!
        req.path = call.argument("path")
        req.miniprogramType = call.argument("type") ?: 0
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleOpenCustomerServiceChat(call: MethodCall, result: Result) {
        val req = WXOpenCustomerServiceChat.Req()
        req.corpId = call.argument("corpId")!!
        req.url = call.argument("url")!!
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleOpenBusinessView(call: MethodCall, result: Result) {
        val req = WXOpenBusinessView.Req()
        req.businessType = call.argument("businessType")!!
        req.query = call.argument("query")
        req.extInfo = call.argument("extInfo")
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handleOpenBusinessWebview(call: MethodCall, result: Result) {
        val req = WXOpenBusinessWebview.Req()
        req.businessType = call.argument("businessType") ?: 0
        req.queryInfo = call.argument("queryInfo")
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun handlePayCall(call: MethodCall, result: Result) {
        val req = PayReq()
        req.appId = call.argument("appId")
        req.partnerId = call.argument("partnerId")
        req.prepayId = call.argument("prepayId")
        req.nonceStr = call.argument("noncestr")
        req.timeStamp = call.argument("timestamp")
        req.packageValue = call.argument("package")
        req.sign = call.argument("sign")
        iwxapi?.sendReq(req)
        result.success(null)
    }

    private fun getShareFilePath(fileUri: String): String {
        val ctx = applicationContext ?: return Uri.parse(fileUri).path ?: ""
        val api = iwxapi ?: return Uri.parse(fileUri).path ?: ""
        if (api.wxAppSupportAPI >= 0x27000D00 && Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            try {
                val providerInfo = ctx.packageManager.getProviderInfo(
                    ComponentName(ctx, WechatFileProvider::class.java),
                    PackageManager.MATCH_DEFAULT_ONLY
                )
                val file = File(Uri.parse(fileUri).path ?: "")
                val shareFileUri = FileProvider.getUriForFile(ctx, providerInfo.authority, file)
                ctx.grantUriPermission("com.tencent.mm", shareFileUri, Intent.FLAG_GRANT_READ_URI_PERMISSION)
                return shareFileUri.toString()
            } catch (e: PackageManager.NameNotFoundException) {
                // ignore
            }
        }
        return Uri.parse(fileUri).path ?: ""
    }
}

File: ios/Classes/WechatBridgePlugin.swift
import Flutter
import UIKit
import WechatOpenSDK

public class WechatBridgePlugin: NSObject, FlutterPlugin, FlutterApplicationLifeCycleDelegate {
    private var channel: FlutterMethodChannel?
    private var isRunning = false
    private var handleInitialWXReqFlag = false
    private var initialWXReqRunnable: (() -> Void)?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "wechat_bridge", binaryMessenger: registrar.messenger())
        let instance = WechatBridgePlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "registerApp":
            registerApp(call: call, result: result)
        case "handleInitialWXReq":
            handleInitialWXReq(call: call, result: result)
        case "isInstalled":
            result(WXApi.isWXAppInstalled())
        case "isSupportApi":
            result(WXApi.isWXAppSupport())
        case "isSupportStateApi":
            result(WXApi.isWXAppSupportStateAPI())
        case "openWechat":
            result(WXApi.openWXApp())
        case "auth":
            handleAuth(call: call, result: result)
        case "startQrauth", "stopQrauth":
            handleQRAuth(call: call, result: result)
        case "openUrl":
            handleOpenUrl(call: call, result: result)
        case "openRankList":
            handleOpenRankList(call: call, result: result)
        case "shareText":
            handleShareText(call: call, result: result)
        case "shareImage", "shareFile", "shareEmoji", "shareMusic", "shareVideo",
             "shareWebpage", "shareMiniProgram":
            handleShareMedia(call: call, result: result)
        case "subscribeMsg":
            handleSubscribeMsg(call: call, result: result)
        case "launchMiniProgram":
            handleLaunchMiniProgram(call: call, result: result)
        case "openCustomerServiceChat":
            handleOpenCustomerServiceChat(call: call, result: result)
        case "openBusinessView":
            handleOpenBusinessView(call: call, result: result)
        case "openBusinessWebview":
            handleOpenBusinessWebview(call: call, result: result)
        case "pay":
            handlePay(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func registerApp(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let appId = args["appId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "appId is required", details: nil))
            return
        }
        let universalLink = args["universalLink"] as? String
        WXApi.registerApp(appId, universalLink: universalLink ?? "")
        isRunning = true
        result(nil)
    }

    private func handleInitialWXReq(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if !handleInitialWXReqFlag {
            handleInitialWXReqFlag = true
            initialWXReqRunnable?()
            initialWXReqRunnable = nil
        }
        // 幂等：第二次及后续调用直接返回成功，避免 PlatformException(FAILED)
        result(nil)
    }

    private func handleAuth(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let scope = args["scope"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "scope is required", details: nil))
            return
        }
        let req = SendAuthReq()
        req.scope = scope
        req.state = args["state"] as? String ?? ""
        let type = args["type"] as? Int ?? 0
        if type == 0 {
            WXApi.send(req) { _ in }
        } else if type == 1 {
            var vc: UIViewController?
            if #available(iOS 13.0, *) {
                vc = UIApplication.shared.connectedScenes
                    .compactMap { $0 as? UIWindowScene }
                    .flatMap { $0.windows }
                    .first { $0.isKeyWindow }?.rootViewController
            } else {
                vc = UIApplication.shared.keyWindow?.rootViewController
            }
            if let vc = vc {
                WXApi.sendAuthReq(req, viewController: vc, delegate: self) { _ in }
            } else {
                // Flutter 场景下可能拿不到 VC，回退到普通方式
                WXApi.send(req) { _ in }
            }
        }
        result(nil)
    }

    private func handleQRAuth(call: FlutterMethodCall, result: @escaping FlutterResult) {
        // 扫码登录需要 WechatAuthSDK，WechatOpenSDK-Full 可能包含
        result(FlutterError(code: "UNIMPLEMENTED", message: "扫码登录请使用 wechat_kit", details: nil))
    }

    private func handleOpenUrl(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let url = args["url"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "url is required", details: nil))
            return
        }
        let req = OpenWebviewReq()
        req.url = url
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handleOpenRankList(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let req = OpenRankListReq()
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handleShareText(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let text = args["text"] as? String,
              let scene = args["scene"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "text and scene required", details: nil))
            return
        }


        let req = SendMessageToWXReq()
        req.bText = true
        req.text = text
        req.scene = Int32(scene)
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handleShareMedia(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let scene = args["scene"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "scene required", details: nil))
            return
        }
        let req = SendMessageToWXReq()
        req.bText = false
        req.scene = Int32(scene)
        let message = WXMediaMessage()
        message.title = args["title"] as? String ?? ""
        message.description = args["description"] as? String ?? ""
        if let thumbData = args["thumbData"] as? FlutterStandardTypedData {
            message.thumbData = thumbData.data
        }

        switch call.method {
        case "shareImage":
            let obj = WXImageObject()
            if let imageData = args["imageData"] as? FlutterStandardTypedData {
                obj.imageData = imageData.data
            } else if let imageUri = args["imageUri"] as? String,
                      let url = URL(string: imageUri),
                      let data = try? Data(contentsOf: url) {
                obj.imageData = data
            }
            message.mediaObject = obj
        case "shareFile":
            let obj = WXFileObject()
            if let fileData = args["fileData"] as? FlutterStandardTypedData {
                obj.fileData = fileData.data
            } else if let fileUri = args["fileUri"] as? String,
                      let url = URL(string: fileUri),
                      let data = try? Data(contentsOf: url) {
                obj.fileData = data
            }
            obj.fileExtension = args["fileExtension"] as? String ?? ""
            message.mediaObject = obj
        case "shareEmoji":
            let obj = WXEmoticonObject()
            if let emojiData = args["emojiData"] as? FlutterStandardTypedData {
                obj.emoticonData = emojiData.data
            } else if let emojiUri = args["emojiUri"] as? String,
                      let url = URL(string: emojiUri),
                      let data = try? Data(contentsOf: url) {
                obj.emoticonData = data
            }
            message.mediaObject = obj
        case "shareMusic":
            let obj = WXMusicObject()
            obj.musicUrl = args["musicUrl"] as? String ?? ""
            obj.musicDataUrl = args["musicDataUrl"] as? String ?? ""
            obj.musicLowBandUrl = args["musicLowBandUrl"] as? String ?? ""
            obj.musicLowBandDataUrl = args["musicLowBandDataUrl"] as? String ?? ""
            message.mediaObject = obj
        case "shareVideo":
            let obj = WXVideoObject()
            obj.videoUrl = args["videoUrl"] as? String ?? ""
            obj.videoLowBandUrl = args["videoLowBandUrl"] as? String ?? ""
            message.mediaObject = obj
        case "shareWebpage":
            let obj = WXWebpageObject()
            obj.webpageUrl = args["webpageUrl"] as? String ?? ""
            message.mediaObject = obj
        case "shareMiniProgram":
            let obj = WXMiniProgramObject()
            obj.webpageUrl = args["webpageUrl"] as? String ?? ""
            obj.userName = args["username"] as? String ?? ""
            obj.path = args["path"] as? String
            if let hdImageData = args["hdImageData"] as? FlutterStandardTypedData {
                message.thumbData = hdImageData.data
            }
            obj.withShareTicket = args["withShareTicket"] as? Bool ?? false
            obj.miniProgramType = WXMiniProgramType(rawValue: UInt(args["type"] as? Int ?? 0)) ?? .release
            obj.disableForward = args["disableForward"] as? Bool ?? false
            message.mediaObject = obj
        default:
            break
        }
        req.message = message
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handleSubscribeMsg(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let scene = args["scene"] as? Int,
              let templateId = args["templateId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "scene and templateId required", details: nil))
            return
        }
        let req = WXSubscribeMsgReq()
        req.scene = UInt32(scene)
        req.templateId = templateId
        req.reserved = args["reserved"] as? String
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handleLaunchMiniProgram(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let userName = args["username"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "username required", details: nil))
            return
        }
        let req = WXLaunchMiniProgramReq()
        req.userName = userName
        req.path = args["path"] as? String
        req.miniProgramType = WXMiniProgramType(rawValue: UInt(args["type"] as? Int ?? 0)) ?? .release
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handleOpenCustomerServiceChat(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let corpId = args["corpId"] as? String,
              let url = args["url"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "corpId and url required", details: nil))
            return
        }
        let req = WXOpenCustomerServiceReq()
        req.corpid = corpId
        req.url = url
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handleOpenBusinessView(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let businessType = args["businessType"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "businessType required", details: nil))
            return
        }
        let req = WXOpenBusinessViewReq()
        req.businessType = businessType
        req.query = args["query"] as? String
        req.extInfo = args["extInfo"] as? String
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handleOpenBusinessWebview(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let businessType = args["businessType"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "businessType required", details: nil))
            return
        }
        let req = WXOpenBusinessWebViewReq()
        req.businessType = UInt32(businessType)
        req.queryInfoDic = args["queryInfo"] as? [AnyHashable: Any]
        WXApi.send(req) { _ in }
        result(nil)
    }

    private func handlePay(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let partnerId = args["partnerId"] as? String,
              let prepayId = args["prepayId"] as? String,
              let nonceStr = args["noncestr"] as? String,
              let timeStamp = args["timestamp"] as? String,
              let package = args["package"] as? String,
              let sign = args["sign"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "pay params required", details: nil))
            return
        }
        let req = PayReq()
        req.partnerId = partnerId
        req.prepayId = prepayId
        req.nonceStr = nonceStr
        req.timeStamp = UInt32(timeStamp) ?? 0
        req.package = package
        req.sign = sign
        WXApi.send(req) { _ in }
        result(nil)
    }

    // MARK: - FlutterApplicationLifeCycleDelegate

    public func application(_ application: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
        return WXApi.handleOpen(url, delegate: self)
    }

    public func application(_ application: UIApplication, continue userActivity: NSUserActivity, restorationHandler: @escaping ([Any]) -> Void) -> Bool {
        return WXApi.handleOpenUniversalLink(userActivity, delegate: self)
    }
}

extension WechatBridgePlugin: WXApiDelegate {
    public func onReq(_ req: BaseReq) {
        var dict: [String: Any] = [:]
        if let launchReq = req as? LaunchFromWXReq {
            let msg = launchReq.message
            dict["messageAction"] = msg.messageAction ?? ""
            dict["messageExt"] = msg.messageExt ?? ""
            dict["lang"] = launchReq.lang ?? ""
            dict["country"] = launchReq.country ?? ""
            if isRunning {
                channel?.invokeMethod("onLaunchFromWXReq", arguments: dict)
            } else {
                initialWXReqRunnable = { [weak self] in
                    self?.channel?.invokeMethod("onLaunchFromWXReq", arguments: dict)
                }
            }
        } else if let showReq = req as? ShowMessageFromWXReq {
            let msg = showReq.message
            dict["messageAction"] = msg.messageAction ?? ""
            dict["messageExt"] = msg.messageExt ?? ""
            dict["lang"] = showReq.lang ?? ""
            dict["country"] = showReq.country ?? ""
            if isRunning {
                channel?.invokeMethod("onShowMessageFromWXReq", arguments: dict)
            } else {
                initialWXReqRunnable = { [weak self] in
                    self?.channel?.invokeMethod("onShowMessageFromWXReq", arguments: dict)
                }
            }
        }
    }

    public func onResp(_ resp: BaseResp) {
        var dict: [String: Any] = [
            "errorCode": resp.errCode,
            "errorMsg": resp.errStr ?? ""
        ]
        switch resp {
        case let authResp as SendAuthResp:
            if resp.errCode == 0 {
                dict["code"] = authResp.code ?? ""
                dict["state"] = authResp.state ?? ""
                dict["lang"] = authResp.lang ?? ""
                dict["country"] = authResp.country ?? ""
            }
            channel?.invokeMethod("onAuthResp", arguments: dict)
        case is OpenWebviewResp:
            channel?.invokeMethod("onOpenUrlResp", arguments: dict)
        case is SendMessageToWXResp:
            channel?.invokeMethod("onShareMsgResp", arguments: dict)
        case let subResp as WXSubscribeMsgResp:
            if resp.errCode == 0 {
                dict["openId"] = subResp.openId ?? ""
                dict["templateId"] = subResp.templateId ?? ""
                dict["scene"] = subResp.scene
                dict["action"] = subResp.action ?? ""
                dict["reserved"] = subResp.reserved ?? ""
            }
            channel?.invokeMethod("onSubscribeMsgResp", arguments: dict)
        case let mpResp as WXLaunchMiniProgramResp:
            if resp.errCode == 0 {
                dict["extMsg"] = mpResp.extMsg ?? ""
            }
            channel?.invokeMethod("onLaunchMiniProgramResp", arguments: dict)
        case is WXOpenCustomerServiceResp:
            channel?.invokeMethod("onOpenCustomerServiceChatResp", arguments: dict)
        case let bvResp as WXOpenBusinessViewResp:
            if resp.errCode == 0 {
                dict["businessType"] = bvResp.businessType ?? ""
                dict["extMsg"] = bvResp.extMsg ?? ""
            }
            channel?.invokeMethod("onOpenBusinessViewResp", arguments: dict)
        case let bwResp as WXOpenBusinessWebViewResp:
            if resp.errCode == 0 {
                dict["businessType"] = bwResp.businessType
                dict["resultInfo"] = bwResp.result ?? ""
            }
            channel?.invokeMethod("onOpenBusinessWebviewResp", arguments: dict)
        case let payResp as PayResp:
            if resp.errCode == 0 {
                dict["returnKey"] = payResp.returnKey ?? ""
            }
            channel?.invokeMethod("onPayResp", arguments: dict)
        default:
            break
        }
    }
}
